src/saveEventsFile.m

```diff
@@ -289,32 +325,55 @@ function printHeaderExtraColumns(logFile)
 
         logFile = checklLogFile('fields', logFile, iEvent, cfg);
 
-        onset = logFile(iEvent).onset;
-        duration = logFile(iEvent).duration;
-        trial_type = logFile(iEvent).trial_type;
-
-        % we skip events with onset or duration that are empty, nan or char
-        if any(cell2mat(cellfun(@isnan, {onset duration}, 'UniformOutput', false))) || ...
-                any(cellfun(@ischar, {onset duration})) || ...
-                any(isempty({onset duration}))
-
-            warning('saveEventsFile:emptyEvent', ...
-                    '\nSkipping saving this event.\n onset: %s \n duration: %s\n', ...
-                    onset, ...
-                    duration);
-
-        else
-
-            printData(logFile(1).fileID, onset, cfg);
-            printData(logFile(1).fileID, duration, cfg);
-            printData(logFile(1).fileID, trial_type, cfg);
+        % if this is _events file, we skip events with onset or duration 
+        % that are empty, nan or char. 
+        if logFile(1,1).isStim==0 
+            
+            onset = logFile(iEvent).onset;
+            duration = logFile(iEvent).duration;
+            trial_type = logFile(iEvent).trial_type;
+
+            if any(cell2mat(cellfun(@isnan, {onset duration}, 'UniformOutput', false))) || ...
+               any(cellfun(@ischar, {onset duration})) || ...
+               any(isempty({onset duration})) 
+
+                warning('saveEventsFile:emptyEvent', ...
+                        '\nSkipping saving this event.\n onset: %s \n duration: %s\n', ...
+                        onset, ...
+                        duration);
+            else
+                printData(logFile(1).fileID, onset, cfg);
+                printData(logFile(1).fileID, duration, cfg);
+                printData(logFile(1).fileID, trial_type, cfg);
+                
+                printExtraColumns(logFile, iEvent, cfg);
 
-            printExtraColumns(logFile, iEvent, cfg);
+                fprintf(logFile(1).fileID, '\n');
+                fprintf(1, '\n');
 
-            fprintf(logFile(1).fileID, '\n');
-            fprintf(1, '\n');
+            end
+            
+        % if this is _stim file, we skip missing events (i.e. events where 
+        % all extra columns have NO values)
+        elseif logFile(1,1).isStim==1
+            
+            skipEvent = false; 
+            namesExtraColumns = returnNamesExtraColumns(logFile);
+            for iExtraColumn = 1:numel(namesExtraColumns)
+                data = logFile(iEvent).(namesExtraColumns{iExtraColumn}); 
+                if isempty(data) || isnan(data) || ( ischar(data) && strcmp(data,'n/a') )
+                    skipEvent = true; 
+                end
+            end
```

Comment:
That seems too restrictive to me to skip saving anything if any of the extracolumns field is empty.

The reason I chose to do it for events is that `onset` and  `duration` are necessary for valid events file (not having `trial_type` is fine though).

Also having `nan` or `n/a` for stim should be fine should not be a reason to skip saving.

So maybe this can be simplified to set `skipEvent` to `true` only if all the ALL the extraColums are empty: what do you think?

Reply:
Here, I think the question is how users will be saving the _stim.tsv file. 

E.g. for the _events.tsv file, I typically want to write the data for each trial in real time, as I go through the experiment, as opposed to collecting it in a bit structure and hoping that matlab doesn't crash so I can successfully write it to the text file at the end of my experiment :D For this, skipping the empty events is crucial (e.g. when I'm writing trial 3, `logFile(1)` and `logFile(2) `would have empty fields for the event columns and hence would be skipped when writing). 

Not sure this is the way the _stim file would be typically written. I guess it doesn't make much sense to write the log about the stimulus "on-the-run", because you know this info in advance? 

In that case, maybe we can just remove the skipping for _stim.tsv altogether?

Reply:
yeah I don't want to make too many assumptions but I think that if indeed there is no data and of it all or it is all nan or n/a then we can skip it.

Reply:
sounds good!